### PR TITLE
design: rook multus design

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

Add the design doc for integrating multus with the ceph backend.

[skip ci]

This PR is really an open discussion, nothing is set in stone and the goal is to find what's best for us as a group.

### CODE REVIEW

**Comment by leseb** [comment]
@Deepthidharwar PTAL and feel free to add any knowledgeable person on the matter. Also, feel free to correct me :). Thanks!

File: design/ceph/multus-network.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  cluster:
+```
+
+Each selector represents a [`NetworkAttachmentDefinition`](https://github.com/intel/multus-cni/blob/master/doc/quickstart.md#storing-a-configuration-as-a-custom-resource) object in Multus.
+At least, one must be provided and by default they will represent:
+
+- `public`: data daemon public network (binds to `public_network` Ceph option).
+- `cluster`: data daemon cluster network (binds to `cluster_network` Ceph option).
+
+If only `public` is set then `cluster` will take the value of `public`.
```
Comment (travisn) [inline design/ceph/multus-network.md]: What if only `cluster` is set? Will `public` use a default?

File: design/ceph/multus-network.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+## Monitors and OSDs implementation challenges
+
+The Ceph monitors only need access to the public network.
+The OSDs needs access to both public and cluster networks.
+
+Monitors requirements so far are the following:
+
+- Predictable IP addresses
+- Keep the same IP address for the life time of the pod (IP should survive a restart)
```
Comment (travisn) [inline design/ceph/multus-network.md]: The IP address should be for the lifetime of the mon deployment. Lifetime of the pod could be shorter if someone deletes the pod.

File: design/ceph/multus-network.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+Pros:
+
+- Pods will get a dedicated IP on a physical network interface
+- No changes required in Ceph, Rook will detect the CIDR via the `NetworkAttachmentDefinition` then populate Ceph's flag `public_network` and `cluster_network`
+
+Cons:
+
+- IP allocation not predictable, we don't know it until the pod is up and running.
+So the detection must append inside the monitor container is running, similarly to what the OSD code does today.
+- Requires drastic changes in the monitor bootstrap code
```
Comment (travisn) [inline design/ceph/multus-network.md]: How can this work at all for mons? If the mon IP changes, the core identity of the mon changes. It seems like we have to change the mons to allow for changing IP addresses, which I didn't think was an option.

File: design/ceph/multus-network.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+Assuming we go with this solution, we might need to change the way the monitors are bootstrapped:
+
+1. let the monitor discovered its own IP based on an interface
+2. once the first mon is bootstrapped, we register its IP in a ConfigMap as well as populating clusterInfo
+3. boot the second mon, look up in clusterInfo for the initial member (if the op dies in the process, we always `CreateOrLoadClusterInfo()` as each startup based on the cm so no worries)
+4. go on and on with the rest of the monitors
+
+TBT: if the pod restarts, it keeps the same IP.
```
Comment (travisn) [inline design/ceph/multus-network.md]: what is TBT? to be tested?

File: design/ceph/multus-network.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+### Second proposal: IPAM type 'DHCP' with service IP
+
+We could and this is pure theory at this point use a IPAM with DHCP along with service IP.
```
Comment (travisn) [inline design/ceph/multus-network.md]: DHCP means IP addresses can change, which leads to the same question of how to deal with mon IPs changing...

File: design/ceph/multus-network.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+This cannot at scale because this means we will have to allocate IPs for **all** the daemon so it's not scalable.
+
+## Monitors and OSDs implementation challenges
+
+The Ceph monitors only need access to the public network.
+The OSDs needs access to both public and cluster networks.
+
+Monitors requirements so far are the following:
+
+- Predictable IP addresses
```
Comment (travisn) [inline design/ceph/multus-network.md]: If multus helps us with the ip address bindings in the pod for all the benefits of security and performance, it seems like the K8s service could still give us the stable endpoint. Is there a reason Multus won't work with the K8s services?

Comment (s1061123) [reply design/ceph/multus-network.md]: For now, multus does not interwork with Kubernetes service object/component (i.e. kube-proxy). In addition, kube-proxy IP service is always in container host, hence some multus interfaces (like SR-IOV interface) send/recv traffic without container host routing/kube-proxy iptables as well.

Multus team is now designing service abstraction for multus. So it is pretty TBD item.

File: design/ceph/multus-network.md (hunk 1 above)
Comment (leseb) [reply design/ceph/multus-network.md]: No.

File: design/ceph/multus-network.md (hunk 4 above)
Comment (leseb) [reply design/ceph/multus-network.md]: Yes.

File: design/ceph/multus-network.md (hunk 6 above)
Comment (travisn) [reply design/ceph/multus-network.md]: @leseb We are blocked from moving to Multus on this issue, or is there another approach we can use for the mons?

File: design/ceph/multus-network.md (hunk 3 above)
Comment (leseb) [reply design/ceph/multus-network.md]: Well, I think there is an aggressive option where failover mons more frequently...

File: design/ceph/multus-network.md (hunk 6 above)
Comment (leseb) [reply design/ceph/multus-network.md]: Yes, we are, not as far as I know. I need to jump into the network plumbing group weekly meeting to ask questions... Probably somewhere after KubeCon.

**Comment by leseb** [comment]
> Independent from the multus design:
> 
> * How about if we move the design docs to the `ceph` folder in a separate PR?

https://github.com/rook/rook/pull/4321

> * I would also think there are a number of places we will need to update links to these design docs.

Indeed.

**Comment by leseb** [comment]
The design has been updated based on the discussion with @dougbtv at KubeCon.

**Comment by BlaineEXE** [comment]
I haven't been privy to all the conversations, but this looks good to me currently.

File: design/ceph/multus-network.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+- IP allocation not predictable, we don't know it until the pod is up and running.
+So the detection must append inside the monitor container is running, similarly to what the OSD code does today.
+- Requires drastic changes in the monitor bootstrap code
+- This adds a DHCP daemon on every part of the cluster and this has proven to be troublesome
+
+Assuming we go with this solution, we might need to change the way the monitors are bootstrapped:
+
+1. let the monitor discovered its own IP based on an interface
+2. once the first mon is bootstrapped, we register its IP in a ConfigMap as well as populating clusterInfo
```
Comment (travisn) [inline design/ceph/multus-network.md]: Can the operator discover this ip address and then update the configmap? Or do we need something in the mon pod to discover it? None of us want to add the rook image back into the mon pods though.

File: design/ceph/multus-network.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+- Pods will get a dedicated IP on a physical network interface
+- No changes required in Ceph, Rook will detect the CIDR via the `NetworkAttachmentDefinition` then populate Ceph's flag `public_network` and `cluster_network`
+
+Cons:
+
+- IP allocation not predictable, we don't know it until the pod is up and running.
+So the detection must append inside the monitor container is running, similarly to what the OSD code does today.
+- Requires drastic changes in the monitor bootstrap code
+- This adds a DHCP daemon on every part of the cluster and this has proven to be troublesome
```
Comment (travisn) [inline design/ceph/multus-network.md]: Ensuring a constant ip address is still the key point to be clear on. If it's using DHCP how can we ensure a stable ip address? Set the expiration on the ip to never expire? Or how can we ensure the ip address doesn't change at some point in the future when the mon pod is restarted?

File: design/ceph/multus-network.md (hunk 7 above)
Comment (leseb) [reply design/ceph/multus-network.md]: We don't know the IP address until the pod is up and running so the operator cannot discover it. And yes, we don't want to use the Rook binary anymore.

File: design/ceph/multus-network.md (hunk 8 above)
Comment (leseb) [reply design/ceph/multus-network.md]: Currently, the DHCP does not support permanent lease AFAIK.

**Comment by leseb** [comment]
@travisn how about we go ahead with this and I'll update it over time if things get to change? Thanks

**Comment by travisn** [comment]
Sounds good to go ahead with the design doc, if you could just clarify that using multus for mons is currently blocked due to the issues mentioned.

File: design/ceph/multus-network.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+If you need a way to assign IP addresses dynamically across your cluster -- Whereabouts is the tool for you. If you've found that you like how the host-local CNI plugin works, but, you need something that works across all the nodes in your cluster (host-local only knows how to assign IPs to pods on the same node) -- Whereabouts is just what you're looking for.
+Whereabouts can be used for both IPv4 & IPv6 addressing.
+
+It is under active development and is not ready but ultimately will allow to:
+
+- Have static IP addresses distributed across the cluster
+- These IPs will survive any deployment restart
+- The allocation will be done by whereabouts and not by Rook
+
+/!\
```
Comment (travisn) [inline design/ceph/multus-network.md]: Is this supposed to be rendered as a warning in markdown? Looks like it is not rendered

Comment (leseb) [reply design/ceph/multus-network.md]: No, just a highlight :)

**Pull request merged**